microservice/challenge-4/src/main/scala/com/byond/Main.scala

```diff
@@ -0,0 +1,135 @@
+import akka.actor.typed.scaladsl.ActorContext
+
+import scala.collection.mutable.ArrayBuffer
+import java.util.Date
+import slick.jdbc.H2Profile.api._
+import slick.jdbc.H2Profile.backend.DatabaseDef
+
+import scala.concurrent.duration._
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.Implicits.global
+
+
+object RootActor {
+  def insertMoods(cats: TableQuery[Cats], db: DatabaseDef, catStates: ArrayBuffer[CatState]): Unit = {    
+    val newRows = catStates.map({ catState =>
+      val CatState(CatId(id), mood, timestamp) = catState
+      (id, Cat.moodToString(mood), timestamp)
+    })
+    val insertAction = DBIO.seq(cats ++= newRows) // Batch inserts are required for performance
+    val insertFuture = db.run(insertAction)
+    Await.result(insertFuture, Duration.Inf)
+  }
+
+  def spawnCats(context: ActorContext[Response], catNumber: Int, catChangeInterval: FiniteDuration, startTimestamp: Long): ArrayBuffer[ActorRef[Command]] = {
+      var children: ArrayBuffer[ActorRef[Command]] = new ArrayBuffer()
+      for (i <- 1 to catNumber) {
+        val child = context.spawn(Cat(CatState(CatId(i), CatMiaw, startTimestamp), context.self, catChangeInterval), "cat" + i.toString())
+        children += child
+      }
+      children
+  }
+
+  def fetchStatistics(cats: TableQuery[Cats], db: DatabaseDef): StatisticsSummary = {
+    val q = cats.map(x => (x.id, x.mood, x.timestamp))
+    val states: Seq[CatState] = Await.result(db.run(q.result), Duration.Inf).flatMap({
+      case (i, m, t) => Cat.moodFromString(m).map(mood => CatState(CatId(i), mood, t))
+    })
+    Stats.summarize(states)
+  }
+
+  def currentTimestamp(): Long = {
+    new Date().getTime() / 1000
+  }
+
+  def apply(catNumber: Int, catChangeInterval: FiniteDuration, refreshInterval: FiniteDuration, modelDuration: FiniteDuration): Behavior[Response] = {
+    val startTimestamp = currentTimestamp()
+    val maxTimestamp = startTimestamp + modelDuration.toSeconds
+
+    // Initialize the in-memory database
+    val db = Database.forConfig("h2mem")
+    val cats = TableQuery[Cats]
+    Await.result(db.run(DBIO.seq(cats.schema.create)), Duration.Inf)
+
+    // Define the behavior of the Root actor:
+    //  - Create and destroy the cat actors
+    //  - Receive messages that represent the mood of cats, and write them to the database by batches
+    //  - Periodically calculate the statistics of the cat moods, from the beginning of the simulation
+    Behaviors.withTimers { timers => 
+      timers.startTimerWithFixedDelay(KillCats(), modelDuration)
+      timers.startTimerWithFixedDelay(RefreshStatistics(), refreshInterval)
+      Behaviors.setup { context =>
+
+        // Start the cats
+        val children = spawnCats(context, catNumber, catChangeInterval, startTimestamp)
+
+        var messageN = 0
+        var catsKilled = false
+        var catStateBuffer: ArrayBuffer[CatState] = new ArrayBuffer()
+        Behaviors.receiveMessage {
```

Comment:
Sorry for my lack of Akka knowledge, but isn't this function run on multiple threads? In that case, how do you handle read/write race condition with the mutable variables passed in?

Reply:
Good question. This would cause a race condition if we created more than one actor of this type. In this case we are safe because there's only one instance of RootActor.

Reply:
@albertaouad Have you seen my question from yesterday?

Reply:
@helkafen apologies, I missed that.
Here's what I was thinking about, over the duration T of the experiment, a cat is on average in mood X for 30% of the duration, and in mood Y for 20% of the duration and in mood Z for 10% of the duration etc...

Reply:
@albertaouad Right. So with this definition, the average is just the count (which I have) divided by the number of samples. There is no possibility to also calculate a variance or a median value, as it was suggested in the README.